### ISSUE TITLE
deploying to Vercel

### ISSUE THREAD

**Issue by headwinds** [lukeburns/now-graphql#4]
When I apply your now.json to my project and attempt to deploy to Vercel - I get this error 

```
Error! The Runtime now-graphql is using nodejs8.10, which was discontinued by an upstream provider. In turn, we have to ask you to upgrade your Runtime to a more recent version or consult the author for more details.
```

**Comment by konsumer** [comment]
There have been some changes to vercel, since I used this, I think. Originally the idea was to be able to put a few files together and it would generate a node wrapper around it to make a graphql service, but I don't think it will quite work like this, now.

Since I don't really have time to troubleshoot the builder-part, I'd recommend just following a no-config route, as this is what I use personally, now.

set your project up like this:

```
package.json
api/index.js
public/index.html
src/*.graphql
src/*.js
```

Where `index.html` is a playground, `api/index.js` is a basic graphql server, and files in `src/` are your typedefs & resolvers. If you add a `build` script to your package.json that assembles the graphql & resolvers into a standard import for `api/index.js` it should all work pretty fast and relatively painlessly.

I am happy to update this project to work that way, as I think the builder idea, although kinda cool, is a bit deprecated and not really maintained.

**Comment by headwinds** [comment]
@konsumer how about I open a PR with an update to the readme about this setup? Can you share your build script?

I bet most devs approach this repo with the same goals - they want an easy, straight-forward approach to host their GQL API on Vercel. 

I think they won't be deterred by a few simple steps until builder's easy button has been updated - and yes agreed its a cool concept.

**Comment by konsumer** [comment]
Here is my fork with all that:
https://github.com/konsumer/now-graphql

Basically, just `npm i`, then mess with files in `src` and `npm run deploy` or run local dev-server with `npm start`

I can make a PR, but we'll have to make sure @lukeburns wants to move in this direction, as it's a big change.

**Comment by konsumer** [comment]
[Here](https://github.com/konsumer/now-graphql/blob/master/build.js) is the actual build script that collects typedefs and merges them, and collects resolvers and import/merges them.

**Comment by konsumer** [comment]
And as a side note, the build step still happens automatically, it's just in your repo instead of an external build-pack. Vercel calls `npm run build` on deploy (and `vercel dev` which is what `npm start` calls.)

**Comment by konsumer** [comment]
The main thing I like about this approach for my other projects is that I can put whatever I want in public/ & api/, and add more build-steps if I want to. It's just more explicit, and makes it easier to combine lots of other little lambdas and reason about how they all fit together. For example, you could put some more little node/go/python/etc scripts in api/, and add more stuff to public/, if you wanted to. It works with nextjs too, if you setup your project in pages/ and `npm i next react react-dom`.

**Comment by headwinds** [comment]
@konsumer yes please go ahead and make the PR - I can provide any feedback as to what I looking for if it helps. 

I'm presenting tomorrow on how to deploy GQL serverless although haven't quite done it yet - still have a couple more hours - no pressure. 

At [work](https://loblawdigital.co/), we use GQL in Google Cloud - I'd like devs to be able to launch their pet GQL projects to help them learn. 

I've been looking at [apollo-server-vercel]( https://apollo-server-vercel.saeris.io/) and [apollo-server-micro](https://www.npmjs.com/package/apollo-server-micro) - I'll be sure mention this repo as well as one to watch!

**Comment by konsumer** [comment]
> @konsumer yes please go ahead and make the PR

Well, that's really up to @lukeburns but either way you can use my fork.

> I've been looking at apollo-server-vercel and apollo-server-micro - I'll be sure mention this repo as well as one to watch!

Yeh, I am using apollo-server-micro in my fork. Not sure I see the benefits of apollo-server-vercel  over it, as vercel uses micro for it's api functions anyway. Seems to me that both of those basically just need a `build` step that uses my script to do the actual merging, so feel free to use it with whatever you like.

**Comment by headwinds** [comment]
@konsumer I tried your fork and was able to get the `hello` world version [hosted on Vercel](https://now-graphql.headwinds1.vercel.app/) which was satisfying but what I really want is to the talk to a restful api so tried to then explore adding a `dataSources` folder with `apollo-datasource-rest`. 

I have this all working in another app run locally but am stumped on the deployment part. In my pet project, I'm using GQL to talk to a python API and a NodeJS API composing the data. 

So would you consider going a bit further in this demo and have it talk to API endpoint?  In addition to `hello.graphql` example, you could add `world.graphql` example that uses apollo-datasource-rest against [some open REST api](https://swapi.dev/) or perhaps you have a different library in mind.

**Comment by headwinds** [comment]
@konsumer got it! thanks so much for your fork and all your help tonight!

![restful](https://user-images.githubusercontent.com/68402/99755321-7d54f180-2ab8-11eb-9865-d596a8085ab8.png)

I was able to get `apollo-datasource-rest` working against [my python api that reads RSS feeds](https://porthole.now.sh/). My demo is ready for tomorrow - so stoked to share this work! 

The only thing I had to do was keep things very flat - your build script doesn't seem to like nested folders so much or at all.

**Comment by headwinds** [comment]
@konsumer this what I mean about the folder structure - take your hello files and put them in a hello folder. The build.js can find them but I get this error 

```
[Error: Query.hello defined in resolvers, but not in schema]
```

If I don't used folders, everything works

**Comment by konsumer** [comment]
I think it's the difference in glob-patterns. Try it now.I also added a param to `hello`, so you can set the name. I also moved all the resolvers & typdefs to `graphql/` to make it clearer how it fits together, and what is boilerplate/tools.

**Comment by headwinds** [comment]
I pulled your latest master and moved my folders into the new graphql folder and presto - it works! thank you!

Just curious on your thoughts about the missing datasources folder. I've only done one graphql project so I don't really know if its required or not. Maybe it's not missing. Where you would put files that want to talk to other APIs?

**Comment by headwinds** [comment]
this is what I've done to add a datascources folder which I'm using to connect to 2 different APIs

![Screen Shot 2020-11-21 at 10 32 26 AM](https://user-images.githubusercontent.com/68402/99880989-e20d6a80-2be4-11eb-8ff6-65eb5126ca09.png)

Do you like this approach or is there a different way? Again, I've only been exposed to this approach so am wondering if there are alternatives...

**Comment by konsumer** [comment]
> Just curious on your thoughts about the missing datasources folder. I

It's not missing. You can add it, if you want, or structure the project however you like. None of these things are a standard, they're just files that get included in a regular vercel service. If you want to use datasources, context,. schemaDirectives,or any other graphql things, you just add it to the service in `api/index.js`, however you want to. I am using a pre-generated schema.js to pull in globs, so it's all computed what will be injected at runtime, which gets around problems with static (the graphql schema) files you might have on lambdas (which vercel runs on) and makes the code go a bit faster (because the import-graph is all done ahead of time, and the typedefs are pre-merged.)

The reason I use `api/index.js` is that it corresponds to `/api` url on vercel. You can put it wherever you want though, for example in `api/graphql.js` (for` /api/graphql` url) Vercel auto-configures things in `api/`, so they should go there, but otherwise, you can name the file whatever you want, within that dir.

> Do you like this approach or is there a different way? Again, I've only been exposed to this approach so am wondering if there are alternatives...

Not sure what you mean by "approach" it's just a dir with files. There needs to be code somewhere that is globbing them and importing them, it's not some automatic thing. You can use my `build.js` as an example for how to do this. Read [the apollo datasource docs](https://www.apollographql.com/docs/apollo-server/data/data-sources/) to understand how this fits together.

This code-example they have on that page shows using `context` & `dataSources`:

```js
const server = new ApolloServer({
  typeDefs,
  resolvers,
  dataSources: () => {
    return {
      moviesAPI: new MoviesAPI(),
      personalizationAPI: new PersonalizationAPI(),
    };
  },
  context: () => {
    return {
      token: 'foo',
    };
  },
});
```

my build script is only creating `typeDefs` and `resolvers`. You'll need to work out whatever you are trying to do from all these examples. Personally, when  I need these extra things, I just add them directly to the service file.

Automatic globs like this are cool for a quick demo, but I find it gets less clear when it's several different folders (like globs other than resolvers & typedefs) For my brain, it tends to be simpler to just add what I want directly to the service-def (like adding an auth-token to `context`, for example), and let the typedef & resolver files generate the schema.
